pdfplumber/page.py

```diff
@@ -115,6 +119,42 @@ def normalize_color(
     return separate_pattern(tuplefied)
 
 
+class PDFPageAggregatorWithMarkedContent(PDFPageAggregator):
+    """Extract layout from a specific page, adding marked-content IDs to
+    objects where found."""
+
+    cur_mcid: Optional[int] = None
+
+    def begin_tag(self, tag: PSLiteral, props: Optional[PDFStackT] = None) -> None:
+        if isinstance(props, dict) and "MCID" in props:
+            self.cur_mcid = props["MCID"]
+        else:
+            self.cur_mcid = None
+
+    def end_tag(self) -> None:
+        self.cur_mcid = None
+
+    def tag_cur_item(self, item_type: Any) -> None:
+        # Implementation Inheritance Considered Harmful
+        cur_obj = self.cur_item._objs[-1]
+        assert isinstance(cur_obj, item_type)
+        # Hackery, which would not be necessary if pdfminer.six supported MCIDs
+        cur_obj.mcid = self.cur_mcid
+
+    def render_char(self, *args, **kwargs) -> float:  # type: ignore
+        adv = super().render_char(*args, **kwargs)
+        self.tag_cur_item(LTChar)
+        return adv
+
+    def render_image(self, *args, **kwargs) -> None:  # type: ignore
```

Comment:
What is the reason for tagging `image` and `char` objects, but not `line`, `rect`, or `curve` objects?

Reply:
Yes, indeed, this was wrong.  I've fixed it to tag the lines/rects/curves with the MCID.